proposals/distribution.md

```diff
@@ -0,0 +1,66 @@
+# Abstract
+
+The Docker registry protocol has become the defacto standard across the container registry world ([https://github.com/docker/distribution/blob/master/docs/spec/api.md](https://github.com/docker/distribution/blob/master/docs/spec/api.md)).
+
+In the OCI, having a solid, common distribution specification with conformance testing will ensure long lasting security and interoperability throughout the container ecosystem.
+
+## Proposal
```

Comment:
One thing I would like to clarify is whether we will actually be making improvements to the spec over time (as per how we did things in the other specifications) or whether we're just going to freeze on the current v2.0 and do nothing much afterwards -- since according to SemVer we'd need to work on v3.0 for that.

Also, it was my understanding that the submission of Docker Distribution as a specification would be something like "any future distribution spec we agree upon will have support for Docker Distribution" -- so as to unblock the distribution problem without locking us into never making significant progress on Distribution.

Reply:
@cyphar I don't think there is anything in this protocol or proposal preventing "significant progress". The specification has been out and in the open since its inception and we have yet to receive any sizable improvement proposals. Most of the perceived limitations of this protocol are limitations in the client implementation that is part of docker.

There are some good changes that we can make in backwards compatible ways:

1. The listing PRs I mentioned here.
1. Stateless transfer. There are only a few small changes required to make this work.
2. p2p already can work on this protocol.

In addition, to this, there is a lot more that can be integrated on top of this protocol as part of client resolution. Most of the issues around naming, signing and mirroring can be taken care of with no API changes.

However, I think it should be clear that the initial specification activity should be about ensuring that the specification matches the current state of the world without breaking compatibility.

Reply:
I think what would be good to see is some clear instructions for the non-expert in terms of writing and working on specifications. Coming from this kind of boat, I have the general sentiment that I want to contribute, and I want to learn about the process to improve expertise in this sort of contribution. I would look for something akin to a CONTRIBUTING.md, or even a simple bullet list in a README that goes through logical steps. Like:

 - Goal: to resolve conflict for technical specifications
 - Procedure:
 1. receiive / link to protocol for discussion
 2. board comments and asks questions as issues
 3. assignment goes to X
 4. issued discussed and resolved, something else...
 5. head maintainer(s) have final sign off on something
 6. publication via... X

I suppose this is a technical standard for reviewing technical standards, haha. For example, a nice parallel is to look at something like [JOSS](http://joss.theoj.org/) that has complete orchestration via a little robot integration. The workflow is clearly defined for the reviewer and reviewee, and the assigned editor. The parties involved fill in the gaps in terms of opening and resolving issues, but there is always clear definition. I'm not sure how something like this might fit for our group, but it would be great to think about to get greater community involvement and discussion.

Reply:
> The workflow is clearly defined for the reviewer and reviewee, and the assigned editor.

project-template (referenced from this proposal) is what the OCI ises for this role.  See especially [here][1] and [here][2].

[1]: https://github.com/opencontainers/project-template/blob/master/CONTRIBUTING.md
[2]: https://github.com/opencontainers/project-template/blob/master/MAINTAINERS_GUIDE.md

Reply:
Thank @wking ! The general roles for a contributor and maintainer are well defined, but I'm wondering about a more specific "hand holding" how to contribute sort of document, even just a bullet list. If it's something that can be observed then I can observe, learn, and write something up. I'm mostly thinking about this to help new contributors such as myself to get into the ropes of the group :)

Reply:
> I'm wondering about a more specific "hand holding" how to contribute sort of document, even just a bullet list.

"How to contribute" is exactly what `CONTRIBUTING.md` tries to cover.  What do you feel us m8ssing?  Perhaps file project-template issues / PRs to avoid drift here?

Reply:
Sure thing! A specific question before closing the discussion here. For this PR, given that we don't have code (nothing to test or critique), and it's a general text draft, what are the criteria we are using for evaluating it, and how do we know when it's ready to go? I think this is likely something that would be obvious for someone working on a lot of these drafts over time, and I'll just catch on, and it would be helpful if someone could jot down a few notes about these questions.

Reply:
> For this PR, given that we don't have code (nothing to test or critique), and it's a general text draft, what are the criteria we are using for evaluating it, and how do we know when it's ready to go?

Ah, *this* repo.  I thought you meant the coming distribution-spec repo.  This repo could use a `CONTRIBUTING.md`, but the voting semantics are covered in the OCI charter (linked from [here][1]).  At some point a TOB member will think it ready and put it up for a TOB vote.  Until then, we're free to suggest improvements.

[1]: https://github.com/opencontainers/tob/tree/93c9b0edefee38921f0c55526564b9b637384004#voting

Reply:
@stevvooe 

Right. Can you expand on these points (we had a short chat a while ago about them, but I don't think I fully understood what the plan was):

> * Stateless transfer. There are only a few small changes required to make this work.
> * p2p already can work on this protocol.

Also, I would still like to have some sort of `.well-known` identifier for registries so that you don't need to host a separate subdomain to have a registry. This would also be a good place to pin future extensions if it turns out that we do want to add something.

Just on this point:

> The specification has been out and in the open since its inception and we have yet to receive any sizable improvement proposals.

The reason for this may be unrelated to whether people want to make changes, or have worthwhile improvements. As you mentioned, a lot of the percieved issues with Distribution are actually because the Docker client doesn't expose those features -- and so people who want to improve distribution may not want to go through improving all layers of the Docker stack to do so.

I guess what I'm saying is "let's not close the door on any improvement discussions", especially since this is now going to be an OCI spec and not a Docker one anymore.

Reply:
@cyphar what sort of language are you looking for to ensure the door is not closed for improvement discussions?  

How about after the interoperability statement adding:
```
This OCI distribution specification will also provide the ecosystem with a means to discuss and schedule extensions to the distribution specification.
```

Reply:
>> The specification has been out and in the open since its inception and we have yet to receive any sizable improvement proposals.
>
> The reason for this may be unrelated to whether people want to make changes, or have worthwhile improvements.
> [...]
> I guess what I'm saying is "let's not close the door on any improvement discussions", especially since this is now going to be an OCI spec and not a Docker one anymore.

To add to what @cyphar stated, I think that the pivot from a Docker spec to an OCI spec changes both the set and focus of the contributors; since the spec is no longer being driven by the Docker product's needs it can attract use-cases that would not have made sense for the Docker product.

@mikebrow That language looks fairly reasonable to me.